Public/Src/Engine/Scheduler/FileMonitoringViolationAnalyzer.cs

```diff
@@ -575,78 +662,137 @@ private Pip TryFindProducer(AbsolutePath producedPath, VersionDisposition versio
+                        isWhitelistedViolation: false,
+                        related: relatedPip,
+                        // we don't have the path of the process that caused the file access violation, so 'blame' the main process (i.e., the current pip) instead
+                        pip.Executable.Path));
+            }
 
-                    // Check there are no source sealed directories containing the write
-                    // Observe that source sealed directories are not allowed above a shared opaque, which means that any source sealed directory
-                    // containing the access will be under the shared opaque (or they will share roots)
-                    var sourceSealedContainer = m_graph.TryGetSealSourceAncestor(access);
-                    if (sourceSealedContainer != DirectoryArtifact.Invalid)
-                    {
-                        var relatedPip = m_graph.GetSealedDirectoryPip(sourceSealedContainer, PipQueryContext.FileMonitoringViolationAnalyzerClassifyAndReportAggregateViolations);
+            // Check if the shared opaque write happens inside a temp directory
+            if (m_graph.TryGetTempDirectoryAncestor(access, out var ownerPip, out var tempPath))
+            {
+                // We found a shared opaque directory that contains a temp directory
+                reportedViolations.Add(
+                    HandleDependencyViolation(
+                        DependencyViolationType.WriteToTempPathInsideSharedOpaque,
+                        AccessLevel.Write,
+                        access,
+                        pip,
+                        isWhitelistedViolation: false,
+                        related: ownerPip,
+                        tempPath));
+            }
+        }
 
-                        // We found a source seal directory that contains the accessed path
-                        reportedViolations.Add(
-                            HandleDependencyViolation(
-                                DependencyViolationType.WriteInSourceSealDirectory,
-                                AccessLevel.Write,
-                                access,
-                                pip,
-                                isWhitelistedViolation: false,
-                                related: relatedPip,
-                                // we don't have the path of the process that caused the file access violation, so 'blame' the main process (i.e., the current pip) instead
-                                pip.Executable.Path));
-                    }
+        private void ReportWriteViolations(
+            Process pip, 
+            List<ReportedViolation> reportedViolations, 
+            IReadOnlyDictionary<FileArtifact, FileMaterializationInfo> outputArtifactsInfo, 
+            AbsolutePath access,
+            [CanBeNull] Dictionary<FileArtifact, (FileMaterializationInfo fileMaterializationInfo, ReportedViolation reportedViolation)> allowedDoubleWriteViolations)

```

Comment:
>Dictionary<FileArtifact, (FileMaterializationInfo fileMaterializationInfo, ReportedViolation reportedViolation)> [](start = 24, length = 112)

This is typed so many times, maybe it is worth considering adding a type alias.

Reply:
Adding aliases to (named) value tuples is only available in >= C#7.0...